### PR TITLE
Graph builder: main graph from the inspection snapshot

### PR DESCRIPTION
**Pull request by jitu5**
**Sub-task:** #2656 · **Parent:** #2265

Second slice of the inspection-adapter backend: the main graph. Stacks on top of the merged Foundation PR (#2689) and targets the **`feat/backend_v2`** integration branch, not `main`.

## Description

Purely **additive**: this PR adds the snapshot graph builder and its parity tests. There is no runtime wiring yet; the builder is validated against the captured legacy-backend baseline but does not yet serve `/api/main`.

- **`integrations/kedro/inspection/graph_builder.py`**: adds `GraphBuilder`, which turns a `ProjectSnapshot` into the main graph response: task, data, and parameter nodes; edges; tag and pipeline lists; and per-node registered-pipeline membership and tags. Membership and tags are aggregated across all registered pipelines, matching the legacy backend. Node IDs use the foundation helpers `_create_task_node_id` and `_create_dataset_node_id`.
- **`integrations/kedro/inspection/__init__.py`**: re-exports `GraphBuilder`.
- **`tests/test_inspection_adapter/test_graph_builder.py`**: validates structural parity against the captured baseline across every registered pipeline view, including node sets by name and type, edge connectivity, top-level tags and pipelines, task display names, and per-node tags and pipeline membership. Modular-pipeline nodes and edges are excluded because they are implemented in a later slice.
- **`tests/test_inspection_adapter/test_graph_builder_edge_cases.py`**: provides focused hermetic coverage for task identity, global tag aggregation, pipeline selection and ordering, source and sink nodes, transcoding, unregistered in-memory datasets, and empty catalog types.

## Deferred to later phases

The required response fields remain valid using empty placeholders:

- **Modular pipelines** (phase 3): `modular_pipelines={}` on the response and `modular_pipelines=None` on nodes.
- **Layers** (phase 4): `layers=[]` on the response and `layer=None` on datasets.
- **Resolved parameter values and node metadata** (phase 5/6): `parameters={}` on task nodes.
- **Runtime wiring**: the builder is not yet routed to serve `/api/main`; that is handled in a later phase.

## Notes

- Parity is compared **structurally, not by literal ID**. Node IDs deliberately changed in D9, so tests compare node names, types, and connectivity rather than hashes.
- `dataset_type` uses the raw catalog type from the snapshot, such as `pandas.CSVDataset`, rather than the legacy backend’s resolved class path. The baseline parity tests therefore do not compare this field. Focused edge-case tests cover raw, unregistered in-memory, and empty dataset types.

## Verification

Validated in the environment with Python 3.14 and Kedro 1.4.0, on top of the merged foundation branch:

- `ruff check`: clean
- `ruff format --check`: clean
- `mypy --config-file package/mypy.ini`: clean
- `pytest package/tests/test_inspection_adapter/`: **75 passed**, 0 failed
- `graph_builder.py`: **100% line coverage**

### CODE REVIEW

File: package/kedro_viz/integrations/kedro/inspection/graph_builder.py [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+
+    def pipeline_ids(self) -> list[str]:
+        """Registered pipeline IDs in this snapshot view (preserves declaration order)."""
+        return list(self._pipelines)
+
+    def build(self, pipeline_id: str | None = None) -> GraphAPIResponse:
+        """Build the graph response for ``pipeline_id`` (default pipeline when ``None``)."""
+        selected = pipeline_id or self.default_pipeline_id()
+        pipeline = self._pipelines[selected]
+
```
Comment (ravi-kumar-pilla) [inline package/kedro_viz/integrations/kedro/inspection/graph_builder.py:147]: **Suggestion:** `build()` indexes `self._pipelines[selected]` directly and will raise an opaque `KeyError` for an unknown `pipeline_id`. Consider validating with `has_pipeline()` first and raising a clearer error (e.g. `ValueError("Invalid pipeline ID: ...")`). The legacy router does this check before building the response; aligning here will make the later `/api/main` wiring phase safer.

File: package/kedro_viz/integrations/kedro/inspection/graph_builder.py [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+    @staticmethod
+    def _add_edge(
+        edges: dict[tuple[str, str], GraphEdgeAPIResponse], source: str, target: str
+    ) -> None:
+        edges.setdefault(
+            (source, target), GraphEdgeAPIResponse(source=source, target=target)
+        )
+
+
+def _display_name(snapshot_name: str, namespace: str | None) -> str:
```
Comment (ravi-kumar-pilla) [inline package/kedro_viz/integrations/kedro/inspection/graph_builder.py:254]: **Suggestion:** `_display_name()` encodes important UX logic (namespace stripping + auto-name hash removal) but is only covered indirectly via baseline parity. A small set of focused unit tests (auto-named, `explicit_diff_func`, no-namespace) would make regressions easier to diagnose than a full baseline diff.

File: package/kedro_viz/integrations/kedro/inspection/graph_builder.py [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+    def has_pipeline(self, pipeline_id: str) -> bool:
+        """Whether ``pipeline_id`` is a registered pipeline in this snapshot view."""
+        return pipeline_id in self._pipelines
+
+    def pipeline_ids(self) -> list[str]:
+        """Registered pipeline IDs in this snapshot view (preserves declaration order)."""
+        return list(self._pipelines)
+
+    def build(self, pipeline_id: str | None = None) -> GraphAPIResponse:
+        """Build the graph response for ``pipeline_id`` (default pipeline when ``None``)."""
```
Comment (ravi-kumar-pilla) [inline package/kedro_viz/integrations/kedro/inspection/graph_builder.py]: **Suggestion:** Public methods like `build()`, `default_pipeline_id()`, and `has_pipeline()` have one-line docstrings but not full Google-style `Args:` / `Returns:` / `Raises:` sections. Worth aligning with the project's API-doc convention before this module becomes part of the public adapter surface.

**Review by ravi-kumar-pilla — APPROVED** [review]
## Kedro PR Review
> Generated with `review-kedro-pr` skill.

### Overview
- **Critical:** 0 | **Suggestions:** 3

### Critical (must fix before merge)
No issues found.

### Suggestions (consider improving)
- [x] `graph_builder.py:L87` — validate `pipeline_id` before dict lookup
- [x] `graph_builder.py:L199` — add focused unit tests for `_display_name`
- [x] `graph_builder.py:L84` — expand public method docstrings to Google style

### Notes
- **General review:** Well-scoped additive slice. Global indexing (pipeline membership + tags) vs per-pipeline rendering matches the live `DataAccessManager` merge-on-shared-node-id behaviour. The transcoding fix (`is_transcoded` flag) is correct.
- **Testing:** Strong coverage — parametrized structural parity across all 6 demo-project pipelines plus hermetic edge cases. 75 inspection-adapter tests passing with 100% line coverage on `graph_builder.py`.
- **CI:** Unit tests, lint, and mypy pass. E2E fails only on Python 3.12 with a flaky `FileNotFoundError` during Behave cleanup — unrelated to this PR.
- **PR checklist:** N/A for core `kedro/` `RELEASE.md`. Deferred placeholders are clearly documented. No backward-compatibility concerns since nothing is routed to `/api/main` yet.
- **API usage:** Correct use of foundation helpers and tag/pipeline aggregation logic.

**Verdict:** Looks good to merge.

File: package/kedro_viz/integrations/kedro/inspection/graph_builder.py (hunk 1 above)
Comment (jitu5) [reply package/kedro_viz/integrations/kedro/inspection/graph_builder.py:147]: Done.

File: package/kedro_viz/integrations/kedro/inspection/graph_builder.py (hunk 3 above)
Comment (jitu5) [reply package/kedro_viz/integrations/kedro/inspection/graph_builder.py]: Done.

File: package/kedro_viz/integrations/kedro/inspection/graph_builder.py (hunk 2 above)
Comment (jitu5) [reply package/kedro_viz/integrations/kedro/inspection/graph_builder.py:254]: added.

File: package/kedro_viz/integrations/kedro/inspection/graph_builder.py [hunk 4]
```diff
@@ -0,0 +1,232 @@
+"""Build the main Kedro-Viz graph response from an inspection snapshot.
+
+Creates task, data and parameter nodes; edges; and global tags and pipeline membership.
```
Comment (ravi-kumar-pilla) [inline package/kedro_viz/integrations/kedro/inspection/graph_builder.py]:
```suggestion
Creates task, data, parameter nodes, edges, global tags and pipeline dictionaries.
```

File: package/kedro_viz/integrations/kedro/inspection/graph_builder.py [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+Modular-pipeline structure, layers and resolved task parameters use empty placeholders. Node IDs
```
Comment (ravi-kumar-pilla) [inline package/kedro_viz/integrations/kedro/inspection/graph_builder.py]: > Modular-pipeline structure, layers and resolved task parameters use empty placeholders.

Lets remove this instead of mentioning it in the docstrings

File: package/kedro_viz/integrations/kedro/inspection/graph_builder.py [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+
+# Match the live backend's get_dataset_type(MemoryDataset()) output for datasets
+# absent from the snapshot catalog.
+MEMORY_DATASET_TYPE = "io.memory_dataset.MemoryDataset"
+
+
+class GraphBuilder:
+    """Build ``GraphAPIResponse`` objects for a project snapshot.
+
+    Pipeline membership and tags are global (aggregated across every registered pipeline), while
```
Comment (ravi-kumar-pilla) [inline package/kedro_viz/integrations/kedro/inspection/graph_builder.py]: lets not use `membership` terminology, it seems a bit odd..wdyt ?

File: package/kedro_viz/integrations/kedro/inspection/graph_builder.py [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+
+    def __init__(self, snapshot: ProjectSnapshot) -> None:
+        self._snapshot = snapshot
+        self._pipelines = {pipeline.name: pipeline for pipeline in snapshot.pipelines}
+        self._task_pipelines: dict[str, set[str]] = defaultdict(set)
+        self._dataset_pipelines: dict[str, set[str]] = defaultdict(set)
+        self._task_tags: dict[str, set[str]] = defaultdict(set)
+        self._dataset_tags: dict[str, set[str]] = defaultdict(set)
+        self._index_nodes()
+        self._tags = self._build_tags()
```
Comment (ravi-kumar-pilla) [inline package/kedro_viz/integrations/kedro/inspection/graph_builder.py]: The `_task_pipelines`, `_dataset_pipelines`, `_task_tags`, and `_dataset_tags` names read like domain objects (TaskNode, DataNode, Tag) rather than lookup tables. These are cross-pipeline aggregation maps keyed by graph identity, not collections of nodes.

Suggested renaming for clarity:

- `_task_pipelines` → `_pipelines_by_task_id`
- `_dataset_pipelines` → `_pipelines_by_dataset_name`
- `_task_tags` → `_tags_by_task_id`
- `_dataset_tags` → `_tags_by_dataset_name`
- `_index_nodes()` → `_aggregate_cross_pipeline_metadata()` (or `_build_cross_pipeline_lookups()`)

The `by_task_id` / `by_dataset_name` suffix makes the key explicit without colliding with `TaskNode` / `DataNode` class names.

File: package/kedro_viz/integrations/kedro/inspection/graph_builder.py [hunk 8]
```diff
@@ -0,0 +1,232 @@
...
+    from kedro.inspection.models import NodeSnapshot, ProjectSnapshot
+
+_AUTO_NAME_RE = re.compile(r"^(?P<func>.+)__[0-9a-f]{8}$")
+
+# Match the live backend's get_dataset_type(MemoryDataset()) output for datasets
+# absent from the snapshot catalog.
+MEMORY_DATASET_TYPE = "io.memory_dataset.MemoryDataset"
+
+
+class GraphBuilder:
```
Comment (ravi-kumar-pilla) [inline package/kedro_viz/integrations/kedro/inspection/graph_builder.py:84]: `GraphBuilder` is doing two distinct jobs:

1. **Global index** - one pass over all pipelines to aggregate per-node `pipelines` and `tags` (the lookup dicts built in `_index_nodes()`)
2. **Per-pipeline render** - walk the selected pipeline and emit nodes/edges for `GraphAPIResponse`

As modular-pipeline and layer logic land in later phases, can we consider extracting (1) into a small `SnapshotGraphIndex` (built once from the snapshot), with `build()` handling only (2). This will make the file manageable and separate concerns.

File: package/kedro_viz/integrations/kedro/inspection/graph_builder.py [hunk 9]
```diff
@@ -0,0 +1,232 @@
...
+            for node in pipeline.nodes:
+                task_id = _create_task_node_id(node.name, node.inputs, node.outputs)
+                self._task_pipelines[task_id].add(pipeline_id)
+                self._task_tags[task_id].update(node.tags)
+                for io in [*node.inputs, *node.outputs]:
+                    stripped = _strip_transcoding(io)
+                    self._dataset_pipelines[stripped].add(pipeline_id)
+                    self._dataset_tags[stripped].update(node.tags)
+
+    def default_pipeline_id(self) -> str:
```
Comment (ravi-kumar-pilla) [inline package/kedro_viz/integrations/kedro/inspection/graph_builder.py:98]: can we put this within build ? I do not see it is used in other places and seems like a small helper which can be inline

File: package/kedro_viz/integrations/kedro/inspection/graph_builder.py [hunk 10]
```diff
@@ -0,0 +1,232 @@
...
+
+        Returns:
+            ``__default__`` when it is registered, otherwise the first registered pipeline ID
+            (declaration order).
+        """
+        if DEFAULT_REGISTERED_PIPELINE_ID in self._pipelines:
+            return DEFAULT_REGISTERED_PIPELINE_ID
+        return next(iter(self._pipelines))
+
+    def has_pipeline(self, pipeline_id: str) -> bool:
```
Comment (ravi-kumar-pilla) [inline package/kedro_viz/integrations/kedro/inspection/graph_builder.py:114]: Lets make this inline as well for now as it is used only in `build`

File: package/kedro_viz/integrations/kedro/inspection/graph_builder.py (hunk 9 above)
Comment (jitu5) [reply package/kedro_viz/integrations/kedro/inspection/graph_builder.py:98]: I’d prefer to keep this small method because callers may need to determine the default pipeline independently of building the graph. The later `/api/main` wiring will use it, and it has focused unit coverage. Let me know if you would still prefer it to be inlined for this phase.

File: package/kedro_viz/integrations/kedro/inspection/graph_builder.py (hunk 10 above)
Comment (jitu5) [reply package/kedro_viz/integrations/kedro/inspection/graph_builder.py:114]: I’d prefer to keep this method because the runtime wiring will need to validate a pipeline ID before calling build(), allowing the API layer to return a clear 404 instead of catching a ValueError. It also keeps that validation in one place and has focused unit coverage.

File: package/kedro_viz/integrations/kedro/inspection/graph_builder.py (hunk 8 above)
Comment (jitu5) [reply package/kedro_viz/integrations/kedro/inspection/graph_builder.py:84]: Done.

File: package/kedro_viz/integrations/kedro/inspection/graph_builder.py (hunk 4 above)
Comment (jitu5) [reply package/kedro_viz/integrations/kedro/inspection/graph_builder.py]: done.

File: package/kedro_viz/integrations/kedro/inspection/graph_builder.py (hunk 5 above)
Comment (jitu5) [reply package/kedro_viz/integrations/kedro/inspection/graph_builder.py]: done.

File: package/kedro_viz/integrations/kedro/inspection/graph_builder.py (hunk 6 above)
Comment (jitu5) [reply package/kedro_viz/integrations/kedro/inspection/graph_builder.py]: Correct, let me update it.

File: package/kedro_viz/integrations/kedro/inspection/graph_builder.py [hunk 11]
```diff
@@ -37,34 +40,59 @@
...
+                task_id = _create_task_node_id(node.name, node.inputs, node.outputs)
+                self._pipelines_by_task_id[task_id].add(pipeline_id)
+                self._tags_by_task_id[task_id].update(node.tags)
+                self._all_tags.update(node.tags)
+                for io in [*node.inputs, *node.outputs]:
+                    stripped = _strip_transcoding(io)
+                    self._pipelines_by_dataset_name[stripped].add(pipeline_id)
+                    self._tags_by_dataset_name[stripped].update(node.tags)
+
+    def task_pipelines(self, task_id: str) -> list[str]:
```
Comment (ravi-kumar-pilla) [inline package/kedro_viz/integrations/kedro/inspection/graph_builder.py]: introducing these names `task_pipelines` in functions again causes a bit of confusion as you addressed earlier. Can we rename these to -`get_pipelines_for_task_id` or `pipelines_by_task_id`

Same for others down here -
```
get_tags_for_task_id, get_pipelines_for_dataset_name, get_tags_for_dataset_name (I prefer these)
or
tags_by_task_id, pipelines_by_dataset_name, tags_by_dataset_name
```

File: package/kedro_viz/integrations/kedro/inspection/graph_builder.py [hunk 12]
```diff
@@ -37,34 +40,59 @@
...
+        return sorted(self._all_tags)
+
+
 class GraphBuilder:
     """Build ``GraphAPIResponse`` objects for a project snapshot.
 
-    Pipeline membership and tags are global (aggregated across every registered pipeline), while
-    only the rendered nodes and edges are scoped to the selected pipeline.
+    Tags and the pipelines each node belongs to are global (aggregated across every registered
+    pipeline by :class:`_SnapshotGraphIndex`), while only the rendered nodes and edges are scoped
```
Comment (ravi-kumar-pilla) [inline package/kedro_viz/integrations/kedro/inspection/graph_builder.py]: I think we do not use `:class:` in our doc strings.

**Comment by ravi-kumar-pilla** [comment]
Hi @jitu5 ,

Thanks for fixing my initial comments. Left some minor comments but code looks good. I will do a re-review tomorrow. Thank you 👍

File: package/kedro_viz/integrations/kedro/inspection/graph_builder.py [hunk 13]
```diff
@@ -0,0 +1,253 @@
...
+class GraphBuilder:
+    """Build ``GraphAPIResponse`` objects for a project snapshot.
+
+    Tags and pipelines associated with each node are aggregated across all registered pipelines,
+    while rendered nodes and edges are scoped to the selected pipeline.
+    """
+
+    def __init__(self, snapshot: ProjectSnapshot) -> None:
+        self._snapshot = snapshot
+        self._pipelines = {pipeline.name: pipeline for pipeline in snapshot.pipelines}
```
Comment (ravi-kumar-pilla) [inline package/kedro_viz/integrations/kedro/inspection/graph_builder.py]: For readability - `self._pipelines_by_id`

File: package/kedro_viz/integrations/kedro/inspection/graph_builder.py [hunk 14]
```diff
@@ -0,0 +1,253 @@
...
+            pipeline_id: The pipeline ID to check.
+
+        Returns:
+            ``True`` if ``pipeline_id`` is a registered pipeline in this snapshot view.
+        """
+        return pipeline_id in self._pipelines
+
+    def pipeline_ids(self) -> list[str]:
+        """Registered pipeline IDs in this snapshot view (preserves declaration order)."""
+        return list(self._pipelines)
```
Comment (ravi-kumar-pilla) [inline package/kedro_viz/integrations/kedro/inspection/graph_builder.py]: For readability - return list(self._pipelines_by_id.keys())  # if renaming the attribute
if not renaming the attribute - return list(self._pipelines.keys()) # just for readability. I know it still returns keys but it is not obvious

File: package/tests/test_inspection_adapter/test_graph_builder.py [hunk 15]
```diff
@@ -0,0 +1,137 @@
...
+) -> None:
+    assert builder.build(pipeline_id).selected_pipeline == expected
+
+
+@pytest.mark.parametrize("pipeline_id", ALL_PIPELINES)
+@pytest.mark.parametrize("node_type", ["task", "data", "parameters"])
+def test_per_node_fields_match_baseline(
+    builder: GraphBuilder, pipeline_id: str, node_type: str
+) -> None:
+    """Per-node pipeline membership and tags match every baseline view."""
```
Comment (ravi-kumar-pilla) [inline package/tests/test_inspection_adapter/test_graph_builder.py]:
```suggestion
    """Per-node pipeline dictionary and tags match every baseline view."""
```

File: package/tests/test_inspection_adapter/test_graph_builder_edge_cases.py [hunk 16]
```diff
@@ -0,0 +1,263 @@
...
+
+
+def _snapshot(
+    pipelines: list[SimpleNamespace],
+    datasets: dict[str, SimpleNamespace] | None = None,
+) -> SimpleNamespace:
+    return SimpleNamespace(pipelines=pipelines, datasets=datasets or {})
+
+
+def test_task_pipeline_membership_uses_task_identity_not_name() -> None:
```
Comment (ravi-kumar-pilla) [inline package/tests/test_inspection_adapter/test_graph_builder_edge_cases.py]:
```suggestion
def test_task_pipelines_use_task_identity_not_name() -> None:
```

File: package/tests/test_inspection_adapter/test_graph_builder_edge_cases.py [hunk 17]
```diff
@@ -0,0 +1,263 @@
...
+            [
+                _pipeline("first_pipe", [_node("a", ["x"], ["y"])]),
+                _pipeline("second_pipe", [_node("b", ["y"], ["z"])]),
+            ]
+        )
+    )
+    assert builder.default_pipeline_id() == "first_pipe"
+
+
+def test_has_pipeline_reports_registered_membership() -> None:
```
Comment (ravi-kumar-pilla) [inline package/tests/test_inspection_adapter/test_graph_builder_edge_cases.py]:
```suggestion
def test_has_pipeline_reports_registered_pipelines() -> None:
```

File: package/tests/test_inspection_adapter/test_graph_builder_edge_cases.py [hunk 18]
```diff
@@ -0,0 +1,263 @@
...
+    ds_node = next(
+        node
+        for node in builder.build("__default__").nodes
+        if node.type == "data" and node.name == "ds"
+    )
+    assert isinstance(ds_node, DataNodeAPIResponse)
+    assert ds_node.dataset_type is None
+
+
+def test_transcoded_variants_create_unique_nodes_and_edges() -> None:
```
Comment (ravi-kumar-pilla) [inline package/tests/test_inspection_adapter/test_graph_builder_edge_cases.py]: `test_transcoded_variants_create_unique_nodes_and_edges` sounds like it asserts separate nodes per ds@pandas / ds@spark. It actually only checks there are no duplicate node IDs or edges (transcoded variants merge via _create_dataset_node_id hashing the base name).

Better name: `test_transcoded_variants_share_one_dataset_node` or `test_transcoded_inputs_do_not_duplicate_nodes_or_edges`.

File: package/kedro_viz/integrations/kedro/inspection/graph_builder.py [hunk 19]
```diff
@@ -0,0 +1,253 @@
...
+    def default_pipeline_id(self) -> str:
+        """Return the default pipeline ID to render.
+
+        Returns:
+            ``__default__`` when it is registered, otherwise the first registered pipeline ID
+            (declaration order).
+        """
+        if DEFAULT_REGISTERED_PIPELINE_ID in self._pipelines:
+            return DEFAULT_REGISTERED_PIPELINE_ID
+        return next(iter(self._pipelines))
```
Comment (ravi-kumar-pilla) [inline package/kedro_viz/integrations/kedro/inspection/graph_builder.py]: if snapshot.pipelines is empty, default_pipeline_id() hits next(iter(self._pipelines)) → StopIteration. Real Kedro projects always have at least one pipeline, so this is low risk; a one-line guard would be nice.

**Review by ravi-kumar-pilla — APPROVED** [review]
I left some readability and minor comments, but overall looks good. Thank you @jitu5

**Review by SajidAlamQB — APPROVED** [review]
Looks good, thank you @jitu5
